src/kms/DeterministicTenantSecurityClient.ts

```diff
@@ -0,0 +1,116 @@
+            .flatMap((deriveResponse) => DetCrypto.batchEncryptField(fields, deriveResponse))
+            .toPromise();
+
+    /**
+     * Decrypt the provided deterministically encrypted field.
+     */
+    decryptField = (encryptedDoc: DeterministicEncryptedField, metadata: FieldMetadata): Promise<DeterministicPlaintextField> =>
+        KmsApi.deriveKey(this.tspDomain, this.apiKey, {[encryptedDoc.secretPath]: [encryptedDoc.derivationPath]}, metadata)
+            .flatMap((deriveResponse) =>
+                DetCrypto.decryptField(encryptedDoc, getDerivedKeys(deriveResponse, encryptedDoc.secretPath, encryptedDoc.derivationPath))
+            )
+            .toPromise();
+
+    /**
+     * Deterministically decrypt a batch of fields using the tenant's KMS that was used for encryption. Supports partial failure and will return both
+     * successfully decrypted fields as well as fields that failed to be decrypted.
+     */
+    decryptFieldBatch = (fields: DeterministicEncryptedFieldCollection, metadata: FieldMetadata): Promise<BatchResult<DeterministicPlaintextField>> =>
+        KmsApi.deriveKey(this.tspDomain, this.apiKey, deterministicCollectionToPathMap(fields), metadata)
+            .flatMap((deriveResponse) => DetCrypto.batchDecryptField(fields, deriveResponse))
+            .toPromise();
+
+    /**
+     * Decrypt the provided deterministically encrypted field and re-encrypt it with the current tenant secret.
+     * This should be called when rotating from one KMS configuration to another.
+     */
+    rotateField = (encryptedDoc: DeterministicEncryptedField, metadata: FieldMetadata): Promise<DeterministicEncryptedField> =>
+        KmsApi.deriveKey(this.tspDomain, this.apiKey, {[encryptedDoc.secretPath]: [encryptedDoc.derivationPath]}, metadata)
+            .flatMap((deriveResponse) => {
+                const derivedKeys = getDerivedKeys(deriveResponse, encryptedDoc.secretPath, encryptedDoc.derivationPath);
+                return DetCrypto.rotateField(encryptedDoc, derivedKeys);
+            })
+            .toPromise();
+
+    /**
+     * Determinally decrypt a batch of fields using the tenant's KMS that was used for encryption, then re-encrypt them with the current tenant secret.
+     * Supports partial failure and will return both successfully re-encrypted fields as well as fields that failed to be re-encrypted.
+     */
+    rotateFieldBatch = (fields: DeterministicEncryptedFieldCollection, metadata: FieldMetadata): Promise<BatchResult<DeterministicEncryptedField>> =>
+        KmsApi.deriveKey(this.tspDomain, this.apiKey, deterministicCollectionToPathMap(fields), metadata)
+            .flatMap((deriveResponse) => DetCrypto.batchRotateField(fields, deriveResponse))
+            .toPromise();
+
+    /**
+     * Deterministically encrypt the provided field with all current and in-rotation secrets for the tenant.
+     * All of the resulting search terms should be used in combination when searching over the column.
```

Comment:
I think you want to use the `field` terminology here. Column is one use case, but `field` is the term we use more broadly.

Reply:
:+1: